File: Codes/0684-redundant-connection_tuoPuPaiXu.py
'''
Author: LetMeFly
Date: 2024-11-02 21:48:09
LastEditors: LetMeFly.xyz
LastEditTime: 2024-11-02 21:52:34
'''
from typing import List

class Solution:
    def union(self, x: int, y: int) -> None:
        self.fa[self.find(x)] = self.find(y)
    
    def find(self, x: int) -> int:
        if self.fa[x] != x:
            self.fa[x] = self.find(self.fa[x])
        return self.fa[x]

    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        self.fa = [i for i in range(len(edges) + 1)]
        for x, y in edges:
            if self.find(x) == self.find(y):
                return [x, y]
            else:
                self.union(x, y)
        return []  # FAKE RETURN

File: Codes/0684-redundant-connection_tuoPuPaiXu.go
/*
 * @Author: LetMeFly
 * @Date: 2024-11-02 21:58:22
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2024-11-02 22:03:24
 */
package main

func find(fa []int, x int) int {
    if fa[x] != x {
        fa[x] = find(fa, fa[x])
    }
    return fa[x]
}

func union(fa []int, x int, y int) {
    fa[find(fa, x)] = find(fa, y)
}

func findRedundantConnection(edges [][]int) []int {
    fa := make([]int, len(edges) + 1)
    for th, _ := range fa {
        fa[th] = th
    }
    for _, edge := range edges {
        if find(fa, edge[0]) == find(fa, edge[1]) {
            return edge
        } else {
            union(fa, edge[0], edge[1])
        }
    }
    return nil
}